### ISSUE TITLE
py312: installing ansible using pip as editable with asdf python is broken

### ISSUE THREAD

**Issue by ssbarnea** [ansible/ansible#81907]
### Summary

When using [asdf](https://github.com/asdf-vm/asdf) or [pyenv]() python and doing editable install, the installation succeeds but it is broken, missing library components.

```
which python
/Users/ssbarnea/.asdf/shims/python
pip install -e .
...
Successfully installed ansible-core-2.17.0.dev0
Reshimming asdf python...
$ ansible --version
Traceback (most recent call last):
  File "/Users/ssbarnea/.asdf/installs/python/3.12.0/bin/ansible", line 5, in <module>
    from ansible.cli.adhoc import main
ModuleNotFoundError: No module named 'ansible.cli.adhoc'
```

This looks like a relatively recent regression but it might not be related to py312 as previously performing editable installs worked fine.

I was not able to spot another python package that fails to install in such way.

### Issue Type

Bug Report

### Component Name

pip

### Ansible Version

```console
$ ansible --version
Traceback (most recent call last):
  File "/Users/ssbarnea/.asdf/installs/python/3.12.0/bin/ansible", line 5, in <module>
    from ansible.cli.adhoc import main
ModuleNotFoundError: No module named 'ansible.cli.adhoc'
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
```

### OS / Environment

Any operating system where python comes from asdf or pyenv, both being tools that use shimming and symlinks to switch python version. Basically `which python` resolves on both of them to a symlink pointing to current python.

### Steps to Reproduce

```yaml (paste below)

```

### Expected Results

Be able to use editable installs.

### Actual Results

```console
broken setup.
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ssbarnea** [comment]
Extra debugging information:
```
$ python
Python 3.12.0 (main, Oct  4 2023, 16:44:35) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from ansible.cli.adhoc import main
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'ansible.cli.adhoc'
>>> import ansible
>>> ansible.__path__
_NamespacePath(['/Users/ssbarnea/.asdf/installs/python/3.12.0/lib/python3.12/site-packages/ansible'])
>>> 
```

```
$ tree /Users/ssbarnea/.asdf/installs/python/3.12.0/lib/python3.12/site-packages/ansible/cli/
/Users/ssbarnea/.asdf/installs/python/3.12.0/lib/python3.12/site-packages/ansible/cli/
├── __pycache__
│   ├── __init__.cpython-312.pyc
│   ├── adhoc.cpython-312.pyc
│   ├── config.cpython-312.pyc
│   ├── doc.cpython-312.pyc
│   ├── galaxy.cpython-312.pyc
│   └── playbook.cpython-312.pyc
└── arguments
    └── __pycache__
        ├── __init__.cpython-312.pyc
        └── option_helpers.cpython-312.pyc
```

This tells me that at least this folder contains some older pre-compiled pyc files but not the modules themselves, which were not installed by pip.

**Comment by ssbarnea** [comment]
I think that I was able to find a way to avoid the bug, I uninstalled the package and removed the ansible folder from site-packages and did an editable install again.

I think that this bug happens only when you have a previous non-editable install existing and you perform an upgrade as editable. The removal of the old version is incomplete (not uninstalling the cache files) and ends up shadowing the editable one.
